posts/bonus-fastify-encapsulation.md

```diff
@@ -0,0 +1,324 @@
+
+async function plugin (instance, opts) {
+  instance.decorate(opts.decoratorName, 42)
+}
+```
+
+Here it should become clearer how encapsulation works and its benefits:
+
+- you can split your source code into different source files/plugins
+- load plugins with different settings (e.g.: connect to many MongoDB connections by registering the plugin twice and changing the connection URL)
+- add hooks to a limited set of routes: the ones defined in the encapsulation context
+
+In fact, if you try to rewrite our examples by moving the `plugin` function to a file,
+you will verify that you can't access the JavaScript clojure anymore to read some external config,
+but you need to rely on the Fastify `instance` argument only!
+
+This is Fastify's secret sauce!
+
+
+### Breaking the encapsulation
+
+You may have seen the [`fastify-plugin`](https://github.com/fastify/fastify-plugin) module that "breaks the encapsulation".
+But why do we need to break the encapsulation?
+
+Let's see the previous example's structure:
+
+[![](https://mermaid.ink/img/pako:eNqNjz0PgjAURf9K82YYdOxgArI4uKgbZXhpn9JIP1IeiUr471ZJnL3TzckZ7p1BB0Mg4ZYw9uLSKC9yXu0pBBZVjIPVyDZ4cfAjo9fUleWu-suqV6tq98EzPVhsupXUP7LNBApwlBxak2fMH0MB9-RIgczVYLorUH7JHk4czk-vQXKaqIApGmRqLOb1DuQVhzFTMpZDOq6_vveWN2Q-TIU)](https://mermaid.live/edit#pako:eNqNjz0PgjAURf9K82YYdOxgArI4uKgbZXhpn9JIP1IeiUr471ZJnL3TzckZ7p1BB0Mg4ZYw9uLSKC9yXu0pBBZVjIPVyDZ4cfAjo9fUleWu-suqV6tq98EzPVhsupXUP7LNBApwlBxak2fMH0MB9-RIgczVYLorUH7JHk4czk-vQXKaqIApGmRqLOb1DuQVhzFTMpZDOq6_vveWN2Q-TIU)
+
+If I would like to access the `ctx1` decorator from the `root` context, I can't do it,
+unless I break the encapsulation by wrapping the plugin function with the `fastify-plugin` module:
+
+```js
+const fp = require('fastify-plugin')
+const app = fastify({ logger: true })
+
+app.decorate('rootDecorator', 42)
+
+app.register(fp(plugin), { decoratorName: 'ctx1' })
+app.register(plugin, { decoratorName: 'ctx2' })
+
+async function plugin (instance, opts) {
+  instance.decorate(opts.decoratorName, 42)
+}
+```
+
+Now the application schema will look like this because we broke the `context 1` context, and the parent swallowed it:
+
+[![](https://mermaid.ink/img/pako:eNqFj7EOwjAMRH8l8twujBmQKF0YWICtYbASQyMaJ0pdCaj67wT6Adx0Or2TfTPY6Ag03DOmXl1aw6ro3Z1iFLVLafAWxUdWBx4F2dL1P1HX22almm4fWegpalN6UEGgHNC7cnD-Egakp0AGdLEO88OA4aVwOEk8v9iCljxRBVNyKNR6LH8G0DccxpKS8xLzcV3wG7J8AKFxRjM)](https://mermaid.live/edit#pako:eNqFj7EOwjAMRH8l8twujBmQKF0YWICtYbASQyMaJ0pdCaj67wT6Adx0Or2TfTPY6Ag03DOmXl1aw6ro3Z1iFLVLafAWxUdWBx4F2dL1P1HX22almm4fWegpalN6UEGgHNC7cnD-Egakp0AGdLEO88OA4aVwOEk8v9iCljxRBVNyKNR6LH8G0DccxpKS8xLzcV3wG7J8AKFxRjM)
+
+Now, the `app` instance has access to the `ctx1` decorator.
+
+Note that every plugin you will install will be wrapped with the `fastify-plugin` module, otherwise
+if the plugin adds a database connection, it will be encapsulated and inaccessible from the parent context!
+
+
+### Why is the Plugin System not easy to understand?
```

Comment:
```suggestion
### Why is the Plugin System so hard to understand?
```

Reply:
I got other reviews here, and I changed from `so hard` to `not easy` because the former was considered as "insulting" to the reader.